ADD basic CMOS driver:

riblanc opened the pull request:
This pull request introduces a basic ~~RTC~~ CMOS driver with the ability to configure its RealTime Clock.

riblanc committed:
ADD basic RTC driver:

Add a basic RTC driver to configure and / or read realtime clock.
This allows us to get the current datetime.

jgiron42 commented on src/drivers/rtc/rtc.zig:
```diff
@@ -0,0 +1,227 @@
+    return ((bcd >> 4) * 10) + (bcd & 0x0F);
+}
+
+fn bin_to_bcd(bin: u8) u8 {
+    return ((bin / 10) << 4) | (bin % 10);
+}
```
I think you should check that bin is less than 100 and otherwise either return an error or panic

jgiron42 commented on src/drivers/rtc/rtc.zig:
```diff
@@ -0,0 +1,227 @@
+    return (read_register_b().data_mode) == .Bcd;
+}
+
+pub fn wait_for_rtc() void {
+    @setRuntimeSafety(true);
+    while (read_register_a().update_in_progress) {}
+}
```
isn't it possible to use the Update-Ended Interrupt and hlt?

riblanc replied:
I don't think so.

With the Update-Ended Interrupt, we only now when the register has completed its update but we can't know if an update is in progress.

The ``wait_for_rtc`` goal, is to ensure there isn't any rtc update in progress, thus it will quit the loop instantly the most of the time.

If I change the logic to wait for the irq to be raised to ensure that if an update takes place it is completed, it could take up to 1 second of waiting each time we want to read the RTC (1 sec if only the CMOS clock interrupt is set, but in practice a lot more interrupt could happens like keyboard / pit etc).
But in any case, halting the CPU could be a big waste of time.

jgiron42 commented on src/drivers/rtc/rtc.zig:
```diff
@@ -0,0 +1,227 @@
+    while (read_register_a().update_in_progress) {}
+}
+
+// Read the current time from the RTC without taking into account the update in progress flag
```
```suggestion
// Read the current time from the RTC
```

riblanc replied:
Whoops, at the beginning, I've split it in 2 functions, one to read the rtc, and one to ensure there isn't any update in progress. I forgot to remove this part from the comment

jgiron42 commented on src/drivers/rtc/rtc.zig:
```diff
@@ -0,0 +1,227 @@
+// Read the current time from the RTC without taking into account the update in progress flag
+pub fn get_time() Datetime {
+    wait_for_rtc();
+    return switch (is_binary_coded_decimal()) {
+        true => .{
+            .seconds = read_register(reg_seconds),
+            .minutes = read_register(reg_minutes),
+            .hours = read_register(reg_hours),
+            .day = read_register(reg_day_of_month),
+            .month = read_register(reg_month),
+            .year = read_register(reg_year),
+            .century = if (reg_century != 0) read_register(reg_century) else DEFAULT_CENTURY,
+        },
+        false => .{
+            .seconds = bcd_to_bin(read_register(reg_seconds)),
+            .minutes = bcd_to_bin(read_register(reg_minutes)),
+            .hours = bcd_to_bin(read_register(reg_hours)),
+            .day = bcd_to_bin(read_register(reg_day_of_month)),
+            .month = bcd_to_bin(read_register(reg_month)),
+            .year = bcd_to_bin(read_register(reg_year)),
+            .century = if (reg_century != 0) bcd_to_bin(read_register(reg_century)) else DEFAULT_CENTURY,
+        },
+    };
```
I think you may have inverted the condition
```suggestion
    return switch (is_binary_coded_decimal()) {
        false => .{
            .seconds = read_register(reg_seconds),
            .minutes = read_register(reg_minutes),
            .hours = read_register(reg_hours),
            .day = read_register(reg_day_of_month),
            .month = read_register(reg_month),
            .year = read_register(reg_year),
            .century = if (reg_century != 0) read_register(reg_century) else DEFAULT_CENTURY,
        },
        true => .{
            .seconds = bcd_to_bin(read_register(reg_seconds)),
            .minutes = bcd_to_bin(read_register(reg_minutes)),
            .hours = bcd_to_bin(read_register(reg_hours)),
            .day = bcd_to_bin(read_register(reg_day_of_month)),
            .month = bcd_to_bin(read_register(reg_month)),
            .year = bcd_to_bin(read_register(reg_year)),
            .century = if (reg_century != 0) bcd_to_bin(read_register(reg_century)) else DEFAULT_CENTURY,
        },
    };
```

riblanc replied:
In fact, I also inverted the DataMode values.
0 correspond to bcd, and 1 correspond to binary.
That's why it worked even tho I inverted condition here I guess :thinking:

Don't blame me I spent like 10 minutes writing this driver :eyes:

jgiron42 commented on src/drivers/rtc/rtc.zig:
```diff
@@ -0,0 +1,227 @@
+}
+
+pub fn set_time(time: Datetime) void {
+    switch (is_binary_coded_decimal()) {
+        true => {
+            write_register(reg_seconds, time.seconds);
+            write_register(reg_minutes, time.minutes);
+            write_register(reg_hours, time.hours);
+            write_register(reg_day_of_month, time.day);
+            write_register(reg_month, time.month);
+            write_register(reg_year, time.year);
+            if (reg_century != 0)
+                write_register(reg_century, time.century);
+        },
+        false => {
+            write_register(reg_seconds, bin_to_bcd(time.seconds));
+            write_register(reg_minutes, bin_to_bcd(time.minutes));
+            write_register(reg_hours, bin_to_bcd(time.hours));
+            write_register(reg_day_of_month, bin_to_bcd(time.day));
+            write_register(reg_month, bin_to_bcd(time.month));
+            write_register(reg_year, bin_to_bcd(time.year));
+            if (reg_century != 0)
+                write_register(reg_century, bin_to_bcd(time.century));
+        },
+    }
```
same thing
```suggestion
    switch (is_binary_coded_decimal()) {
        false => {
            write_register(reg_seconds, time.seconds);
            write_register(reg_minutes, time.minutes);
            write_register(reg_hours, time.hours);
            write_register(reg_day_of_month, time.day);
            write_register(reg_month, time.month);
            write_register(reg_year, time.year);
            if (reg_century != 0)
                write_register(reg_century, time.century);
        },
        true => {
            write_register(reg_seconds, bin_to_bcd(time.seconds));
            write_register(reg_minutes, bin_to_bcd(time.minutes));
            write_register(reg_hours, bin_to_bcd(time.hours));
            write_register(reg_day_of_month, bin_to_bcd(time.day));
            write_register(reg_month, bin_to_bcd(time.month));
            write_register(reg_year, bin_to_bcd(time.year));
            if (reg_century != 0)
                write_register(reg_century, bin_to_bcd(time.century));
        },
    }
```

jgiron42 commented on src/drivers/rtc/rtc.zig:
```diff
@@ -0,0 +1,227 @@
+}
+
+// EXPERIMENTAL: get timestamp from RTC in seconds since 1970/01/01 00:00:00 UTC
+pub fn get_timestamp() u32 {
+    const days_in_month = [_]u32{ 0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
+
+    const time = get_time();
+
+    const years_since_1970 = (@as(u32, time.century) * 100 + @as(u32, time.year)) - 1970;
+    const leap_years = (years_since_1970 + 2) / 4 - (years_since_1970 + 70) / 100 + (years_since_1970 + 370) / 400;
+    const normal_years = years_since_1970 - leap_years;
+    var timestamp = leap_years * 31622400 + normal_years * 31536000;
+
+    var days: u32 = 0;
+    for (0..time.month) |month| {
+        days += if (month == 2 and is_leap_year(1970 + years_since_1970)) 29 else days_in_month[month];
+    }
+    days += @as(u32, time.day - 1);
+
+    timestamp += days * 86400;
+    timestamp += @as(u32, time.hours) * 3600;
+    timestamp += @as(u32, time.minutes) * 60;
+    timestamp += @as(u32, time.seconds);
+
+    return timestamp;
+}
```
I rarely say that but this would deserve a small unit test (and maybe you should separate the datetime to timestamp logic from the get_time(), for example with something like that: `get_time().to_timestamp()`)

riblanc replied:
It's the reason why I marked it as "experimental".
It was a simple test, and is actually not intended to be used yet.
It could lead to many errors like overflow etc,
I will probably remove it sooner or later.

riblanc replied:
In fact, I'm not really sure if I wanna push this experimental thing with my PR tho :thinking:

riblanc commented on src/drivers/rtc/rtc.zig:
```diff
@@ -0,0 +1,227 @@
+};
+
+const DataMode = enum(u1) {
+    Raw = 0,
+    Bcd = 1,
```
```suggestion
    Bcd = 0,
    Bin = 1,
```

riblanc committed:
Apply suggestions from code review

Co-authored-by: Joachim Giron <59100853+jgiron42@users.noreply.github.com>

riblanc committed:
UPDATE PIC driver:

Add enable_slave / try_disable_slave functions:
- enable_slave simply enable the pic_slave irq.
- disable_slave disable the pic_slave irq only if no slave irq is enabled.

riblanc committed:
UPDATE RTC driver:

Rename RTC driver for CMOS driver as RTC is a part of the CMOS chipset.
Add bin_to_bcd protection as requested by @jgiron42 in PR #186.
Add a RegisterF struct and a function to read it, describing the CMOS's 'Shutdown status' infos.
Add functions to enable update_ended, alarm and periodic RTC interrupt.

riblanc commented:
I marked this PR as draft since it rely on #190
I will change its status back when the IRQ fixes will be merged on main, and when dev/task will be up-to-date with main
